Source/Core/Core/HW/DVDThread.cpp

```diff
@@ -0,0 +1,194 @@
+
+static u64 s_dvd_offset;
+static u32 s_output_address;
+static u32 s_length;
+static bool s_decrypt;
+
+// Used to notify emulated software after executing command.
+// Pointers don't work with savestates, so CoreTiming events are used instead
+static int s_callback_event_type;
+
+// The following time variables are only used for logging
+static u64 s_realtime_started_us;
+static u64 s_realtime_done_us;
+
+void Start()
+{
+	s_finish_read = CoreTiming::RegisterEvent("FinishReadDVDThread", FinishRead);
+	_assert_(!s_dvd_thread.joinable());
+	s_dvd_thread = std::thread(DVDThread);
+}
+
+void Stop()
+{
+	_assert_(s_dvd_thread.joinable());
+
+	// The DVD thread will return if s_DVD_thread_exiting
+	// is set when it starts working
+	s_dvd_thread_exiting.Set();
+	s_dvd_thread_start_working.Set();
+
+	s_dvd_thread.join();
+
+	s_dvd_thread_exiting.Clear();
+}
+
+void DoState(PointerWrap &p)
+{
+	WaitUntilIdle();
+
+	// TODO: Savestates can be smaller if s_DVD_buffer is not saved
+	p.Do(s_dvd_buffer);
+	p.Do(s_time_read_started);
+	p.Do(s_dvd_success);
+
+	p.Do(s_dvd_offset);
+	p.Do(s_output_address);
+	p.Do(s_length);
+	p.Do(s_decrypt);
+	p.Do(s_callback_event_type);
+
+	// s_realtime_started_us and s_realtime_done_us aren't savestated
+	// because they rely on the current system's time.
+	// This means that loading a savestate might cause
+	// incorrect times to be logged once.
+}
+
+void WaitUntilIdle()
+{
+	_assert_(Core::IsCPUThread());
+
+	// Wait until DVD thread isn't working
+	s_dvd_thread_done_working.Wait();
+
+	// Set the event again so that we still know that the DVD thread isn't working
+	s_dvd_thread_done_working.Set();
+}
+
+void StartRead(u64 dvd_offset, u32 output_address, u32 length, bool decrypt,
+               int callback_event_type, int ticks_until_completion)
+{
+	_assert_(Core::IsCPUThread());
+
+	s_dvd_thread_done_working.Wait();
+
+	s_dvd_offset = dvd_offset;
+	s_output_address = output_address;
+	s_length = length;
+	s_decrypt = decrypt;
+	s_callback_event_type = callback_event_type;
+
+	s_time_read_started = CoreTiming::GetTicks();
+	s_realtime_started_us = Common::Timer::GetTimeUs();
+
+	s_dvd_thread_start_working.Set();
+
+	CoreTiming::ScheduleEvent(ticks_until_completion, s_finish_read);
+}
+
+static void FinishRead(u64 userdata, int cyclesLate)
+{
+	WaitUntilIdle();
+
+	DEBUG_LOG(DVDINTERFACE, "Disc has been read. Real time: %" PRIu64 " us. "
+	          "Real time including delay: %" PRIu64 " us. Emulated time including delay: %" PRIu64 " us.",
+	          s_realtime_done_us - s_realtime_started_us,
+	          Common::Timer::GetTimeUs() - s_realtime_started_us,
+	          (CoreTiming::GetTicks() - s_time_read_started) / (SystemTimers::GetTicksPerSecond() / 1000 / 1000));
+
+	while (!s_dvd_success)
+	{
+		// Asking "Do you want to retry?" would be more user friendly,
+		// but if panic alerts are turned off, the answer is automatically Yes,
+		// and we don't want an infinite loop where ReadFromDisc always fails...
```

Comment:
The case I was thinking of is someone storing games on external media and accidentally unplugging it. It doesn't happen often, but giving the user a chance to recover is better than desyncing and very likely making the game crash. Boot not caring about return values is an argument for fixing Boot, not for not having this.

Because adding a retry feature doesn't really have anything to do with making disc reads asynchronous and because there are issues with it, I will drop it from this PR. I might end up submitting it as a separate PR later in the future if I manage to make it cleaner (or soon if there's someone who's interested in this feature).